### PR TITLE
Refactor ConfigParser

### PR DESCRIPTION
**Pull request by leokoppel**
Prereq for #234

Refactor ConfigParser. The primary motivation is to allow defining parsing functions in other modules, so wave_utils does not depend on OpenCV, or any future classes we want to parse.

Basically this is a wrapper around yaml-cpp's `convert` mechanism. However, by request I kept the interface to ConfigParser the same, so other classes still use `addParam`, `load(file)`, etc.

Note that despite the templates, the parsing functions for Matrix, etc. are still compiled into the library. Only the same types are still supported.

- Remove huge enum and switch statements of types
- Use templated Param class
- Define enum for return codes instead of int

#### Pre-Merge Checklist:
- [x] Code is documented in doxygen format
- [x] Code has automated tests
- [x] Zero compiler warnings
- [x] Formatted with `clang-format`

### CODE REVIEW

File: wave_utils/include/wave/utils/config.hpp [hunk 1]
```diff
@@ -131,93 +128,125 @@ class ConfigParser {
...
+ *   cols: 3
+ *   data: [1.1, 2.2, 3.3,
+ *          4.4, 5.5, 6.6,
+ *          7.7, 8.8, 9.9]
+ * ```
+ *
+
+ */
+template <>
+struct convert<cv::Mat> {
```
Comment (chutsu) [inline wave_utils/include/wave/utils/config.hpp]: Was this not supposed to be in the vision module?

File: wave_utils/include/wave/utils/config.hpp [hunk 2]
```diff
@@ -25,57 +25,54 @@ namespace wave {
...
+    KeyError = -2,       ///< key not found, parameter is not optional
+    ConversionError = -3 /**< Invalid value for conversion
+                           * e.g. wrong-size list for vector, missing
+                           * key 'rows', 'cols' or 'data' (for matrix) */
+};
+
+/** Base class representing a parameter to be parsed in the yaml file */
+struct ConfigParamBase {
+    ConfigParamBase(std::string key, bool optional)
+        : key{std::move(key)}, optional{optional} {};
```
Comment (chutsu) [inline wave_utils/include/wave/utils/config.hpp]: why move? typo?

File: wave_utils/include/wave/utils/config.hpp [hunk 3]
```diff
@@ -131,93 +128,125 @@ class ConfigParser {
...
+    ConfigStatus getYamlNode(std::string key, YAML::Node &node);
+
+    /** Check whether a key is present in the yaml file */
+    ConfigStatus checkKey(std::string key, bool optional);
+
+    /** Load yaml param primitive, array, vector or matrix. */
+    ConfigStatus loadParam(const ConfigParamBase &param);
+
+    /** Load yaml file at `config_file`. */
+    ConfigStatus load(std::string config_file);
```
Comment (chutsu) [inline wave_utils/include/wave/utils/config.hpp]: `const std::string &` maybe, I know I haven't been putting `const` everywhere but maybe we should?

File: wave_utils/include/wave/utils/config.hpp (hunk 2 above)
Comment (leokoppel) [reply wave_utils/include/wave/utils/config.hpp]: See https://stackoverflow.com/questions/7592630/is-pass-by-value-a-reasonable-default-in-c11
https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html

File: wave_utils/include/wave/utils/config.hpp (hunk 1 above)
Comment (leokoppel) [reply wave_utils/include/wave/utils/config.hpp]: Separate PR #237

File: wave_utils/include/wave/utils/config.hpp (hunk 3 above)
Comment (leokoppel) [reply wave_utils/include/wave/utils/config.hpp]: will do

File: wave_utils/include/wave/utils/config.hpp (hunk 2 above)
Comment (chutsu) [reply wave_utils/include/wave/utils/config.hpp]: This just hurts my head.

**Comment by leokoppel** [comment]
OK, I added some reasoning to the issue. One of our projects does currently have an issue with opencv  which I can tell you about offline.

**Pull request merged**